CTR Nonce Reuse demo

AleX04Nov opened the pull request:
# CTR Nonce Reuse demo

This demo is a little game that utilises technical issues of reusing the nonce in CTR mode,
just hit:
```bash
$ cargo run
```
You will see a window with 3 tabs:
- main tab with game. It has 4 eavesdropped packets. Those are ciphertexts. One of them is the ciphertext of the FLAG value. Also, there is a hint with the plaintext of one of 3 eavesdropped packets.
- xor tool tab. It helps in XORing two different hex values;
- hex encoder/decoder. It helps to convert text into ASCII and vice versa.

## CTR Nonce Reuse demo

Every launch of this demo generates a unique key and init vector. Also, the button `RESTART` changes the game’s values (like the key, nonce, flag, and eavesdropped packets).
To submit the flag: press the `SUBMIT` button or hit an `ENTER` while focusing on the flag textbox.

**Have fun!**

AleX04Nov committed:
init

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+const EAVESDROPPED_MSG_COUNT: usize = 3;
+
+pub struct CtrGame {
+    rng: ThreadRng,
```
There is no need of storing the rng in the game. We always can use `random::thread_rng()` inplace. It will create a new rng seeded by the operating system. Furthermore, storing the rng in the `CtrGame` complicates the usage of the `CtrGame`, because any time we want to generate random numbers, we have to use `&mut self`, which could be easily avoided.

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+pub struct CtrGame {
+    rng: ThreadRng,
+
+    //aes_ctr: Aes128Ctr64LE,
```
Commented lines of code is generally a bad sign. We always have git for storing the old code :)

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+        let right_random_bytes: [u8; 4] = self.rng.gen();
+        let mut flag_str = String::new();
+
+        write!(
```
Creating a string and then immediately writing to it could be easily replace just by `format!`: 
```suggestion
        let flag_str = format!(
```

In this case, there would be no need for `.expect` because it's done implicitly by the `format!`.

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+            .plaintexts_vec
+            .choose_multiple(&mut self.rng, count)
+            .collect();
+        self.known_plaintext = (*chosen_plaintext.choose(&mut self.rng).expect("a")).clone();
```
This star expression with the `clone()` can be replaced by `to_vec()`:
```suggestion
        self.known_plaintext = chosen_plaintext.choose(&mut self.rng).expect("non-empty").to_vec();
```

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    pub fn submit_flag(&self, flag_request: String) -> bool {
+        flag_request.into_bytes().eq(&self.flag_plaintext)
```
The `into_bytes` call creates a new vector, making an allocation. We can avoid that by using `as_bytes()`, which creates a reference to the underlying array:
```suggestion
        flag_request.as_bytes() == self.flag_plaintext
```

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    pub fn hex_to_text(data: String) -> String {
+        let decoded = hex::decode(data);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        let decoded = decoded.expect("We have already handled the error");
+
+        let decoded = String::from_utf8(decoded);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        decoded.expect("We have already handled the error")
```
There is a pattern in this code, where you first check for an error and then unwrap the result if it's ok. This is how the Result-like type would look in the language without pattern matching. But the Rust is not such language :)
For that operation you could use either `match` or `if let`:
```rust
let decoded = match decoded {
    Ok(decoded) => decoded,
    Err(_) => String::from("[unprintable]")
};
```

```rust
let decoded = if let Ok(decoded) {
    decoded
} else {
    String::from("[unprintable]")
}
```

Or just use the `Result::unwrap_or_else`:
```rust
let decoded = decoded.unwrap_or_else(|| String::from("[unprintable]"));
```

P.S. this is not the only place with this pattern.

---

But in this case, you can use helpers in the `Result` and `Option` structs For example:
 ```suggestion
        hex::decode(data)
            .ok()
            .and_then(|data| String::from_utf8(data).ok())
            .unwrap_or_else(|| String::from("[unprintable]"))
```

What each function does:
- [Result::ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok):
  Converts `Ok(data)` into `Some(data)`, and `Err(err)` into `None`.
- [Option::and_then](https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then)
  > Returns [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None) if the option is [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None), otherwise calls f with the wrapped value and returns the result.
- [Option::unwrap_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else):
  > Returns the contained [Some](https://doc.rust-lang.org/std/option/enum.Option.html#variant.Some) value or computes it from a closure.
  
  Note that we could use just the [Option::unwrap_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or), but it would create a new string in any case (even if the result is ok), which makes an allocation. And we could avoid it with the `unwrap_or_else` which will make the allocation only if the result is None.

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,196 @@
+        if data.len() < 2 {
+            return xor_result;
+        }
+        let mut xor_result = data[0].clone();
```
To prevent allocations, we can just pop the last vector in the `data` :)
```suggestion
        let mut xor_result = data.pop().expect("length is checked");
```

AleX04Nov replied:
But `data` is not mutable in this function. I thought that it was a better approach not to mutate incoming variables. In case we would use this function in other places, where we need to store incoming array intact.

G1gg1L3s replied:
If this function accepted `data: &[Vec<u8>]` then yes - the data is immutable and we have no ownership of it. 
But, this functions accepts entire `Vec<Vec<u8>>` by value. This means that the function _owns_ the vector. It means, that the function will at least deallocate it. And if you own the data, no one can prevent you from mutating it :)
In this case this will be useful, because it will save one allocation.

G1gg1L3s reviewed (commented):
I can't say much about the ui code, because the framework is unfamiliar for me and the code looks ok (except for the commented lines).

Overall, nice job! I really like the way it looks :)

AleX04Nov committed:
removing `rng` from CtrGame structure

CtrGame:
- removing `rng` from `CtrGame` structure.

AleX04Nov committed:
removed commented lines of code

CtrGame:
- removed commented lines of code.

AleX04Nov committed:
using `format!` instead of `write!` in flag generator

CtrGame:
- using using `format!` instead of `write!` in flag generation method
- removed `use std::fmt::Write` dependency

AleX04Nov committed:
substituting star expression with `clone` with `to_vec` method

CtrGame:
- substituting star expression with `clone` with `to_vec` method;

AleX04Nov committed:
changed flag evaluation. Removed allocation

CtrGame:
- changed flag evaluation in the `submit_flag` method. Removed allocation.

AleX04Nov committed:
replaced is_err pattern with other patterns

CtrGame:
- replaced `is_err()` pattern in `hex_to_text` and `xor_hex` methods with other patterns.

AleX04Nov committed:
move padding into `main_column` from `Container` builder

Tabs:
- move padding into `main_column` from `Container` and `Scrollable` builder.

AleX04Nov committed:
add scrolling to all tabs

Tabs:
- added scrolling to all tabs.

AleX04Nov committed:
remove unnecessary comment from main

Main:
- removed unnecessary comment.

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -134,6 +129,8 @@ impl CtrGame {
     }
 
     /*
+    Dead Code
+    But we can leave it to show how CTR mode is decrypting
```
If you want to show how to decrypt it, you can leave the code uncommented and make it `pub` :)
The compiler will check that it's indeed a valid code. So, if you continue developing your demo, you will have no chance of breaking it at some point.

AleX04Nov replied:
![image](https://user-images.githubusercontent.com/35816182/175341795-d836a9ee-19d1-4cc1-b05f-5052e93d2c3c.png)
nope(99
compiler still throws warnings

G1gg1L3s replied:
hmm, that's because the `ctr_game` module is not public. But you can use the `#[allow(dead_code)]` attribute:
```rust
#[allow(dead_code)]
fn decrypt_bytes(&self, ciphertext_vec: &[u8]) -> Vec<u8> {
    self.encrypt_bytes(ciphertext_vec)
}
```
Just don't forget to comment why this attribute exists and it will be ok :)

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,183 @@
+                br#"{  "UserType": "UsualUser",     "Name": "Alex", "Surname": "M"}"#.to_vec(),
+                br#"{  "UserType": "UsualUser",     "Name": "Tom",  "Surname": "S"}"#.to_vec(),
+                br#"{  "UserType": "Administrator", "Name": "Alex", "Surname": "S"}"#.to_vec(),
+                //br#"{ "Heartbeat": "Alive",         "Since": "1653924358" }"#.to_vec(),
```
Is it intentionally commented?

G1gg1L3s commented on ctr-nonce-reuse/src/ctr_game_tab.rs:
```diff
@@ -0,0 +1,247 @@
+                        ),
+                )
+                .max_width(300)
+                //.width(Length::Shrink)
```
Leftover :)

G1gg1L3s commented on ctr-nonce-reuse/src/hex_encoder_decoder.rs:
```diff
@@ -0,0 +1,128 @@
+
+    fn tab_label(&self) -> TabLabel {
+        TabLabel::Text(String::from("HEX ENC/DEX"))
+        //TabLabel::IconText(Icon::User.into(), self.title())
```
And here, and in other places :))

AleX04Nov committed:
remove new mem allocation in `xor_vec`

CtrGame:
- remove new mem allocation in the `xor_vec` method.

G1gg1L3s reviewed (commented):
By the way, forgot to ask, how is this GUI framework portable? Will it run on windows or mac?

AleX04Nov committed:
remove unnecessary code comments from the whole project

Whole Demo:
- removed unnecessary code comments from the whole project.

AleX04Nov committed:
make `ctr_game` module public

Main:
- made `ctr_game` module public.

AleX04Nov commented:
> By the way, forgot to ask, how is this GUI framework portable? Will it run on windows or mac?

![image](https://user-images.githubusercontent.com/35816182/175349547-9632fc51-9cdd-4337-8942-8e78fd9e33f8.png)
@G1gg1L3s Yes, it would run with the same UI graphic design as on mac. I have not tested it on Linux yet.

AleX04Nov committed:
update README run command

README:
- updated run command

AleX04Nov committed:
correcting misspelling in victory modal

CtrGame:
- corrected misspelling in victory modal.

G1gg1L3s reviewed (approved):
Nice job, @AleX04Nov!

> I have not tested it on Linux yet.

At least it works on my linux machine, so I believe it's ok :)

vixentael commented on ctr-nonce-reuse/README.md:
```diff
@@ -0,0 +1,19 @@
+- xor tool tab. It helps in XORing two different hex values;
+- hex encoder/decoder. It helps to convert text into ASCII and vice versa.
+
+## CTR Nonce Reuse demo
```
```suggestion
```

vixentael commented on ctr-nonce-reuse/README.md:
```diff
@@ -0,0 +1,19 @@
+
+Every launch of this demo generates a unique key and init vector. Also, the button `RESTART` changes the game’s values (like the key, nonce, flag, and eavesdropped packets).
+To submit the flag: press the `SUBMIT` button or hit an `ENTER` while focusing on the flag textbox.
+
```
please put instructions on what exactly the user is supposed to do :)

smth like "The goal of the game is  ... To get the FLAG, you should do ... "

AleX04Nov committed:
change the random bytes count on the flag from 8 to 4

CtrGame:
- change the random bytes count on the flag from 8 to 4 to shrink the flag string size.

AleX04Nov committed:
change available plaintexts

CtrGame:
- change available plaintexts with different quotations, etc;

AleX04Nov committed:
Update README add sections

README:
- add `HOW TO PLAY` section;
- add `Hint` section;
- add `Step-by-step` section;

AleX04Nov committed:
running cargo fmt

vixentael commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,210 @@
+                br#"We shall fight on the seas of Azov and Black Sea."#.to_vec(),
+                br#"We shall fight on the landing grounds (Hostomel)"#.to_vec(),
+                // Nazis
+                br#"Wollt ihr den totalen Krieg? (c) Putler"#.to_vec(),
```
```suggestion
```

vixentael commented on ctr-nonce-reuse/src/ctr_game.rs:
```diff
@@ -0,0 +1,210 @@
+                br#"Free Ukraine, Donate on u24.gov.ua !"#.to_vec(),
+                br#"Russian Warship - Go F Yourself"#.to_vec(),
+                // https://youtu.be/2Cv6nFwDGvA?t=12
+                br#"Guys, don't take up a defensive position... (UAF to Nazis)"#.to_vec(),
+                br#"Better just to surrender now... (UAF to Nazis)"#.to_vec(),
+                br#"You are all F'ed... (UAF to Nazis)"#.to_vec(),
+                br#"You're on my land, and there you'll feed worms (UAF to Nazis)"#.to_vec(),
```
```suggestion
                br#"Guys, don't take up a defensive position... (UAF to RUS)"#.to_vec(),
                br#"Better just to surrender now... (UAF to RUS)"#.to_vec(),
                br#"You are all F'ed... (UAF to RUS)"#.to_vec(),
                br#"You're on my land, and there you'll feed worms (UAF to RUS)"#.to_vec(),
```

vixentael reviewed (commented):
wording suggestions

AleX04Nov committed:
update plaintext phrases list

CtrGame:
- update the plaintext phrases list.
